src/sundomeigest/ArnI/sundomeigest_arni.c

```diff
@@ -0,0 +1,483 @@
+    if (retval != 0)
+    {
+      if (retval < 0) { return SUN_ERR_DEE_ATIMES_FAIL_UNREC; }
+      else { return SUN_ERR_DEE_ATIMES_FAIL_REC; }
+    }
+
+    SUNCheckCall(SUNModifiedGS(ArnI_CONTENT(DEE)->V, ArnI_CONTENT(DEE)->Hes,
+                               i + 1, ArnI_CONTENT(DEE)->krydim,
+                               &(ArnI_CONTENT(DEE)->Hes[i + 1][i])));
+
+    /* Unitize the computed orthogonal vector */
+    N_VScale(SUN_RCONST(1.0) / ArnI_CONTENT(DEE)->Hes[i + 1][i],
+             ArnI_CONTENT(DEE)->V[i + 1], ArnI_CONTENT(DEE)->V[i + 1]);
+    SUNCheckLastErr();
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDomEigEstimate_ArnI(SUNDomEigEstimator DEE, sunrealtype* lambdaR,
+                                  sunrealtype* lambdaI)
+{
+  SUNFunctionBegin(DEE->sunctx);
+
+  SUNAssert(lambdaR, SUN_ERR_ARG_CORRUPT);
+  SUNAssert(lambdaI, SUN_ERR_ARG_CORRUPT);
+  SUNAssert(ArnI_CONTENT(DEE)->ATimes, SUN_ERR_DEE_NULL_ATIMES);
+  SUNAssert(ArnI_CONTENT(DEE)->V, SUN_ERR_ARG_CORRUPT);
+  SUNAssert(ArnI_CONTENT(DEE)->q, SUN_ERR_ARG_CORRUPT);
+  SUNAssert(ArnI_CONTENT(DEE)->Hes, SUN_ERR_DEE_NULL_HES);
+
+  sunindextype n = ArnI_CONTENT(DEE)->krydim;
+
+  /* Reshape the Hessenberg matrix as an input vector for the LAPACK dgeev_ function */
+  sunindextype i, j, k = 0;
+  for (i = 0; i < n; i++)
+  {
+    for (j = 0; j < n; j++)
+    {
+      ArnI_CONTENT(DEE)->LAPACK_A[k] = ArnI_CONTENT(DEE)->Hes[i][j];
+      k++;
+    }
+  }
+
+  char jobvl = 'N'; // Do not compute left eigenvectors
+  char jobvr = 'N'; // Do not compute right eigenvectors
+
+  /* Call LAPACK's dgeev function
+      return info values refer to
+    = 0:  successful exit
+    < 0:  if info = -i, the i-th argument had an illegal value.
+    > 0:  if info = i, the QR algorithm failed to compute all the
+          eigenvalues, and no eigenvectors have been computed;
+          elements i+1:N of LAPACK_wr and LAPACK_wi contain
+          eigenvalues which have converged.
+  */
+  sunindextype lda  = n;
+  sunindextype ldvl = n;
+  sunindextype ldvr = n;
+  sunindextype info;
+  sunindextype lwork = 4 * n;
+  xgeev_f77(&jobvl, &jobvr, &n, ArnI_CONTENT(DEE)->LAPACK_A, &lda,
+            ArnI_CONTENT(DEE)->LAPACK_wr, ArnI_CONTENT(DEE)->LAPACK_wi, NULL,
+            &ldvl, NULL, &ldvr, ArnI_CONTENT(DEE)->LAPACK_work, &lwork, &info);
+
+  if (info != 0)
+  {
+    // printf(DEE_LAPACK_FAIL, info); // Need to use the error handler
+
+    return SUN_ERR_DEE_LAPACK_FAIL;
+  }
+
+  /* order the eigenvalues by their magnitude */
+  for (i = 0; i < n; i++)
+  {
+    ArnI_CONTENT(DEE)->LAPACK_arr[i][0] = ArnI_CONTENT(DEE)->LAPACK_wr[i];
+    ArnI_CONTENT(DEE)->LAPACK_arr[i][1] = ArnI_CONTENT(DEE)->LAPACK_wi[i];
+  }
+
+  /* Sort the array using qsort */
+  qsort(ArnI_CONTENT(DEE)->LAPACK_arr, n,
+        sizeof(ArnI_CONTENT(DEE)->LAPACK_arr[0]), sundomeigest_Compare);
+
+  // alternatively we can return a vector of all computed dom_eigs (up to krydim)
+  // TODO: Get opinions
```

Comment:
Can this comment be removed?

Reply:
Since the power iteration can only return one dominant eigenvalue, and the base class is designed to just provide one, then this routine should also only return one.  

However, it could be an added benefit to users if you store all of these, and then provide a separate "Get" routine to retrieve all of the estimated eigenvalues from the last "Estimate" call.